File: training/src/timbre_graph_lab/patchmap.py
"""The patch MAP: a 2-D surface whose every point is a good sound.

The 1-D tour has a structural flaw the numbers made plain: it must find one
long path through a k-nearest-neighbour graph, and longest-path is NP-hard, so
it strands itself. Measured 2026-08-03, bass kept 40 spread anchors inside a
37-node connected component and the tour visited **13** of them. Two thirds of
a screened, loudness-normalized, lens-validated pool went unused because it did
not happen to lie on one walk.

A map needs no path. Every kept anchor becomes a point in the plane, and the
control is a position rather than a distance along a route — so all 40 are
reachable, and there are infinitely many blends between them instead of 19
fixed hops.

## Why it does not just become mush

"Can we cheat a bit — I don't want the user to be able to go too far from a
patch, we want the sounds as predictably good as possible" (user, 2026-08-03).

So the blend is deliberately NOT a smooth average over the whole plane. Weights
are inverse distance raised to a high power, which makes the surface a **soft
Voronoi**: across most of its area one anchor dominates so completely that you
are, to the ear, standing on that validated patch. The blending happens in
narrow bands near the boundaries between neighbours. Turn `sharpness` up and
the map approaches "twenty discrete patches with short crossfades"; turn it
down and it approaches a continuous wash. The default sits near the former.

That also answers "travel through the patches, not near them": at an anchor's
own coordinates its weight is 1 and the reproduction is exact.
"""

from __future__ import annotations

import numpy as np

# Neighbours considered at any point. Beyond ~4 the extra weights are
# vanishingly small at usable sharpness, and each one costs a parameter blend.
MAP_NEIGHBOURS = 4
# Inverse-distance exponent — the "cheat" knob. Higher = more Voronoi-like.
#
# Measured over a 40-anchor map, fraction of the surface where one anchor wins
# outright (so the sound IS that validated patch), and the mean dominance of
# the nearest anchor elsewhere:
#
#     sharpness   exact patch   mean top weight
#         2.0          5%            0.53        a wash — everything is a blend
#         4.0         24%            0.70
#         8.0         52%            0.84
#        12.0         66%            0.89
#        20.0         77%            0.93        <- default
#
# Raised from 12 to 20 on 2026-08-05 for SAFETY, not taste. The lurches that
# survived pruning all lived in hybrid territory — a blend of two safe filter
# settings can resonate where neither does — so shrinking that territory is a
# more honest fix than sampling it more densely and hoping.
#
# 20 keeps ~three quarters of the map standing on a checked patch while leaving
# real hybrid territory in the seams, which is the stated goal: "I don't want
# the user to be able to go too far from a patch". Shipped IN the artifact so it
# can be retuned by ear without a code change.
MAP_SHARPNESS = 20.0
# Above this dominance the nearest anchor simply wins, so most of the surface
# reproduces a validated patch EXACTLY rather than to within rounding.
MAP_SNAP = 0.9
_EPS = 1e-9


def embed_2d(z: np.ndarray) -> np.ndarray:
    """Lay descriptor rows out in the unit square. Deterministic. Pure.

    PCA keeps the two directions the pool actually varies along, so patches
    that sound alike land near each other and the map reads as a space rather
    than a scatter. The sign of a principal component is arbitrary, so it is
    pinned (largest-|loading| entry made positive) — otherwise the same corpus
    could produce a mirrored map on a re-run and every stored coordinate would
    silently change meaning.
    """
    n = len(z)
    if n == 0:
        return np.zeros((0, 2))
    if n == 1:
        return np.array([[0.5, 0.5]])
    x = np.asarray(z, dtype=np.float64)
    x = x - x.mean(axis=0)
    # economy SVD: rows are observations, so V rows are the components
    _, _, vt = np.linalg.svd(x, full_matrices=False)
    comps = vt[:2] if len(vt) >= 2 else np.vstack([vt, np.zeros_like(vt[0])])
    for i in range(comps.shape[0]):
        if comps[i][np.argmax(np.abs(comps[i]))] < 0:
            comps[i] = -comps[i]
    xy = x @ comps.T

    # normalise each axis into [0,1]; a degenerate axis collapses to centre
    out = np.zeros_like(xy)
    for j in range(2):
        lo, hi = float(xy[:, j].min()), float(xy[:, j].max())
        out[:, j] = 0.5 if hi - lo < _EPS else (xy[:, j] - lo) / (hi - lo)
    return out


def blend_weights(
    points: np.ndarray,
    xy: tuple[float, float] | np.ndarray,
    k: int = MAP_NEIGHBOURS,
    sharpness: float = MAP_SHARPNESS,
    snap: float = MAP_SNAP,
) -> tuple[np.ndarray, np.ndarray]:
    """(indices, weights) of the anchors contributing at `xy`. Pure.

    This is the whole runtime of the map, and the TypeScript panel mirrors it
    exactly — any change here must be mirrored there or the artifact stops
    describing what the user hears.
    """
    n = len(points)
    if n == 0:
        return np.zeros(0, dtype=int), np.zeros(0)
    p = np.asarray(xy, dtype=np.float64)
    d = np.linalg.norm(np.asarray(points, dtype=np.float64) - p, axis=1)

    idx = np.argsort(d, kind="stable")[: min(k, n)]
    dn = d[idx]
    if dn[0] < _EPS:                      # standing exactly on an anchor
        w = np.zeros(len(idx))
        w[0] = 1.0
        return idx, w

    w = 1.0 / np.power(dn, sharpness)
    w = w / w.sum()
    if w[0] >= snap:                      # one anchor dominates — let it win
        w = np.zeros(len(idx))
        w[0] = 1.0
    return idx, w


def params_at_xy(
    points: np.ndarray,
    snapshots: np.ndarray,
    xy: tuple[float, float] | np.ndarray,
    k: int = MAP_NEIGHBOURS,
    sharpness: float = MAP_SHARPNESS,
    snap: float = MAP_SNAP,
) -> np.ndarray:
    """The parameter vector at a point on the map. Pure."""
    idx, w = blend_weights(points, xy, k=k, sharpness=sharpness, snap=snap)
    if len(idx) == 0:
        return np.zeros(0)
    return (np.asarray(snapshots, dtype=np.float64)[idx] * w[:, None]).sum(axis=0)


# Mirrors CONFIDENCE_FLOOR in TimbreGraphPanel.tsx. The panel plays a position
# at this fraction of full when it has no confidence, scaling to full when one
# validated patch dominates — so a safety gate must measure `rms * this` to
# hear what the user hears.
CONFIDENCE_FLOOR = 0.15


def level_for_confidence(confidence: float) -> float:
    c = min(1.0, max(0.0, float(confidence)))
    return CONFIDENCE_FLOOR + (1.0 - CONFIDENCE_FLOOR) * c


def grid_points(n: int) -> list[tuple[float, float]]:
    """An n x n sample of the unit square, for build-time validation."""
    if n <= 1:
        return [(0.5, 0.5)]
    step = 1.0 / (n - 1)
    return [(i * step, j * step) for i in range(n) for j in range(n)]

File: src/patchmap.ts
/**
 * The X/Y patch map runtime — a direct mirror of the lab's `patchmap.py`.
 *
 * Any change here must be mirrored there (and vice versa), because the build
 * VALIDATES the surface by rendering exactly what these functions produce. If
 * the two drift, the artifact stops describing what the user hears — the
 * failure mode that made a dial sweep painful before it was caught.
 *
 * The blend is deliberately not a smooth average. Weights are inverse distance
 * raised to a high power, which makes the surface a SOFT VORONOI: across most
 * of its area one anchor dominates so completely that you are standing on that
 * validated patch, and the blending happens in narrow bands at the boundaries
 * between neighbours. Measured over a 40-anchor map, `sharpness` 12 leaves
 * about two thirds of the surface reproducing a checked patch exactly.
 */

export interface MapPoint {
  preset_id: string;
  name: string;
  fxp_path: string;
  /** Position in the unit square. */
  xy: [number, number];
}

/**
 * One WORLD for a role: a start patch plus the map of everything reachable
 * through it.
 *
 * The lens supplies the oscillators, filter types and routing that every point
 * on this map is heard through, and the runtime can never write those — so a
 * map is not "the instrument", it is one structure's worth of it. Exhaust a
 * map and the way onward is a different lens, not more dragging.
 */
export interface MapLens {
  lens: { preset_id: string; name: string; category: string };
  param_names: string[];
  points: MapPoint[];
  /** Absolute parameter values at each point, in `param_names` order. */
  snapshots: number[][];
  sharpness: number;
  neighbours: number;
  snap: number;
}

export interface MapRole {
  role: string;
  track_id?: string;
  lenses: MapLens[];
  declined: boolean;
}

export interface MapGraph {
  version: string;
  roles: Record<string, MapRole>;
  quality?: Record<string, unknown>;
}

/**
 * The lens at `index`, WRAPPING.
 *
 * Wrapping rather than clamping because the index is shared across roles that
 * have different numbers of viable lenses, and because a layered group spreads
 * its members across the list — with six members and three lenses, clamping
 * would pile members 3-5 onto the last one while wrapping keeps them spread.
 */
export function lensAt(role: MapRole, index: number): MapLens | null {
  const n = role.lenses?.length ?? 0;
  if (n === 0) return null;
  return role.lenses[((Math.floor(index) % n) + n) % n];
}

const EPS = 1e-9;

/**
 * Indices and weights of the anchors contributing at `xy`.
 *
 * Exported for tests: the snap behaviour and the convexity of the weights are
 * the two properties the whole design rests on.
 */
export function blendWeights(
  points: ReadonlyArray<readonly [number, number]>,
  xy: readonly [number, number],
  neighbours = 4,
  sharpness = 20,
  snap = 0.9,
): { idx: number[]; w: number[] } {
  if (points.length === 0) return { idx: [], w: [] };

  const d = points.map((p, i) => ({
    i,
    d: Math.hypot(p[0] - xy[0], p[1] - xy[1]),
  }));
  d.sort((a, b) => (a.d === b.d ? a.i - b.i : a.d - b.d));
  const near = d.slice(0, Math.min(neighbours, d.length));

  // standing exactly on an anchor: it IS that patch, no arithmetic
  if (near[0].d < EPS) {
    return { idx: near.map((n) => n.i), w: near.map((_, k) => (k === 0 ? 1 : 0)) };
  }

  const raw = near.map((n) => 1 / Math.pow(n.d, sharpness));
  const total = raw.reduce((a, b) => a + b, 0);
  const w = raw.map((v) => v / total);

  // one anchor dominates — let it win outright, so most of the map
  // reproduces a validated patch EXACTLY rather than to within rounding
  if (w[0] >= snap) {
    return { idx: near.map((n) => n.i), w: near.map((_, k) => (k === 0 ? 1 : 0)) };
  }
  return { idx: near.map((n) => n.i), w };
}

/**
 * How certain a position is — the dominance of its nearest anchor, 0..1.
 *
 * 1 means "this IS a validated patch"; lower means the sound is a hybrid of
 * several, which is still inside the convex hull of checked configurations but
 * has not itself been rendered. Used to sit uncertain layers back in the mix.
 */
export function confidenceAt(
  role: Pick<MapLens, 'points' | 'sharpness' | 'neighbours' | 'snap'>,
  xy: readonly [number, number],
): number {
  const pts = role.points.map((p) => p.xy as readonly [number, number]);
  const { w } = blendWeights(pts, xy, role.neighbours, role.sharpness, role.snap);
  return w.length ? w[0] : 1;
}

/**
 * The map point nearest `xy` — where a gesture resolves when it is released.
 */
export function nearestPoint(
  role: Pick<MapLens, 'points'>,
  xy: readonly [number, number],
): readonly [number, number] | null {
  let best: readonly [number, number] | null = null;
  let bd = Infinity;
  for (const p of role.points) {
    const d = Math.hypot(p.xy[0] - xy[0], p.xy[1] - xy[1]);
    if (d < bd) { bd = d; best = p.xy as readonly [number, number]; }
  }
  return best;
}

/** The parameter vector at a point on the map. */
export function paramsAtXY(
  role: Pick<MapLens, 'points' | 'snapshots' | 'sharpness' | 'neighbours' | 'snap'>,
  xy: readonly [number, number],
): number[] {
  const pts = role.points.map((p) => p.xy as readonly [number, number]);
  const { idx, w } = blendWeights(
    pts, xy, role.neighbours, role.sharpness, role.snap,
  );
  if (idx.length === 0) return [];
  const width = role.snapshots[idx[0]]?.length ?? 0;
  const out = new Array<number>(width).fill(0);
  idx.forEach((anchor, k) => {
    const snap = role.snapshots[anchor];
    if (!snap || w[k] === 0) return;
    for (let j = 0; j < width; j++) out[j] += (snap[j] ?? 0) * w[k];
  });
  return out;
}

/**
 * Indices of every parameter the map moves ANYWHERE.
 *
 * Written at every position including where the delta is zero — see the
 * matching note on the tour's `movedParamIndices`. A parameter that is not
 * sent keeps whatever the previous position left it at, which silently makes
 * the control path-dependent.
 */
export function movedParamIndices(snapshots: readonly number[][]): number[] {
  const start = snapshots[0] ?? [];
  const out: number[] = [];
  start.forEach((_, i) => {
    if (snapshots.some((s) => Math.abs((s[i] ?? 0) - (start[i] ?? 0)) > 1e-5)) {
      out.push(i);
    }
  });
  return out;
}
